src/sagas/import-project.saga.test.js

```diff
@@ -0,0 +1,280 @@
+          [dialog, 'showErrorBox'],
+          'Unsupported project type',
+          "Looks like the project you're trying to import isn't supported. Unfortunately, Guppy only supports projects created with create-react-app or Gatsby"
+        )
+      );
+      expect(saga.next().done).toBe(true);
+    });
+
+    it('should handle any other errors', () => {
+      const error = new Error('unknown error');
+      const saga = handleImportError(error);
+      expect(saga.next().value).toEqual(call([console, 'error'], error));
+      expect(saga.next().value).toEqual(
+        call(
+          [dialog, 'showErrorBox'],
+          'Unknown error',
+          'An unknown error has occurred. Sorry about that! Details have been printed to the console.'
+        )
+      );
+      expect(saga.next().done).toBe(true);
+    });
+  });
+
+  describe('importProject saga', () => {
+    const startAction = { path: 'path/to/project' };
+    let json;
+    beforeEach(() => {
+      json = {
+        name: 'example',
+        version: '0.1.0',
+        private: true,
+        dependencies: {
+          react: '^16.4.1',
+          'react-dom': '^16.4.1',
+          'react-image-magnify': '^2.7.0',
+          'react-router': '^4.3.1',
+          'react-router-dom': '^4.3.1',
+          'react-scripts': '1.1.4',
+        },
+        scripts: {
+          start: 'react-scripts start',
+          build: 'react-scripts build',
+          test: 'react-scripts test --env=jsdom',
+          eject: 'react-scripts eject',
+        },
+      };
+    });
+
+    it('should throw error if project already exist', () => {
+      const saga = importProject(startAction);
+      const expectedError = new Error('project-name-already-exists');
+      expect(saga.next().value).toEqual(
+        call(loadPackageJson, 'path/to/project')
+      );
+      expect(saga.next(json).value).toEqual(
+        select(getInternalProjectById, 'example')
+      );
+      expect(saga.next({ name: 'example' }).value).toEqual(
+        call(handleImportError, expectedError)
+      );
+      expect(saga.next().value).toEqual(put(importExistingProjectError()));
+      expect(saga.next().done).toBe(true);
+    });
+
+    it('should throw error if project have unsupported type', () => {
+      const saga = importProject(startAction);
+      const expectedError = new Error('unsupported-project-type');
+      expect(saga.next().value).toEqual(
+        call(loadPackageJson, 'path/to/project')
+      );
+      expect(saga.next(json).value).toEqual(
+        select(getInternalProjectById, 'example')
+      );
+      expect(saga.next().value).toEqual(call(inferProjectType, json));
+      expect(saga.next(null).value).toEqual(
+        call(handleImportError, expectedError)
+      );
+      expect(saga.next().value).toEqual(put(importExistingProjectError()));
+      expect(saga.next().done).toBe(true);
+    });
+
+    it('should import project', () => {
+      global.Date = { now: () => 1532809641976 };
```

Comment:
ah, right, the `createdAt` is done impurely within the saga, same as it was within the middleware.

This solution worries me, though (it won't be "unmocked" when this test file finishes, right? So this could affect subsequent tests). We could have an `afterAll` that restores it, but I wonder if there's a better way...

Maybe `importProject` could take a second optional argument, `getCurrentTime`, which could default to `Date.now`. In the test you could provide a mock function like `() => 1532809641976`.

What do you think?

Reply:
You're right. I can replace it with: 
```javascript
const spyOnDate = jest.spyOn(Date, 'now');
spyOnDate.mockReturnValue(1532809641976);

// test

spyOnDate.mockRestore()
```
in this case mocked value will affect only one testcase